src/test/java/com/squareup/javapoet/TypeNameTest.java

```diff
@@ -41,6 +69,31 @@
     TypeName.get(recursiveEnum.getGenericParameterTypes()[0]);
   }
 
+  @Test
+  public void innerClassInGenericType() throws Exception {
+    Method genericStringInner = getClass().getDeclaredMethod("testGenericStringInner");
+    TypeName.get(genericStringInner.getReturnType());
+    TypeName.get(genericStringInner.getGenericReturnType());
+    assertNotEquals(TypeName.get(genericStringInner.getGenericReturnType()),
+        TypeName.get(getClass().getDeclaredMethod("testGenericIntInner").getGenericReturnType()));
+  }
+
+  @Test
+  public void innerGenericInGenericType() throws Exception {
+    Method genericStringInner = getClass().getDeclaredMethod("testGenericInnerString");
+    TypeName.get(genericStringInner.getReturnType());
+    TypeName.get(genericStringInner.getGenericReturnType());
+    assertNotEquals(TypeName.get(genericStringInner.getGenericReturnType()),
+        TypeName.get(getClass().getDeclaredMethod("testGenericInnerInt").getGenericReturnType()));
+  }
+
+  @Test
+  public void innerStaticInGenericType() throws Exception {
+    Method staticInGeneric = getClass().getDeclaredMethod("testNestedNonGeneric");
+    TypeName.get(staticInGeneric.getReturnType());
+    TypeName.get(staticInGeneric.getGenericReturnType());
```

Comment:
I was going to comment how strange it is to see side-effect free statements being used to implicitly test behavior, but it looks like there's other test methods doing the same thing already. We should change these to assert _something_, but I'll leave that for a follow-up to fix everything.